ganttproject/src/main/java/net/sourceforge/ganttproject/chart/UnboundedViewportTaskSceneBuilder.kt

```diff
@@ -0,0 +1,184 @@
+import biz.ganttproject.core.chart.scene.BarChartConnector
+import biz.ganttproject.core.chart.scene.gantt.Connector
+import biz.ganttproject.core.chart.scene.gantt.DependencySceneBuilder
+import biz.ganttproject.core.chart.scene.gantt.TaskActivitySceneBuilder
+import biz.ganttproject.core.chart.scene.gantt.TaskLabelSceneBuilder
+import biz.ganttproject.core.option.DefaultEnumerationOption
+import biz.ganttproject.core.time.TimeDuration
+import biz.ganttproject.core.time.TimeDurationImpl
+import biz.ganttproject.core.time.impl.GPTimeUnitStack
+import net.sourceforge.ganttproject.GanttProjectImpl
+import net.sourceforge.ganttproject.task.dependency.TaskDependencyConstraint
+import net.sourceforge.ganttproject.task.dependency.TaskDependency.Hardness
+import java.awt.Color
+import java.awt.Dimension
+import java.awt.Rectangle
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class UnboundedViewportTaskSceneBuilder(
+  private val input: TaskActivitySceneBuilder.ChartApi,
+  private val logger: ((String) -> Unit)? = null
+) {
+  fun build(tasks: List<Task>, canvas: Canvas = Canvas()): Canvas {
+    val allTasks = tasks.flatMap { it.flatten() }
+    paintTasks(allTasks, canvas)
+    paintDependencies(allTasks, canvas)
+    return canvas
+  }
+
+  private fun paintTasks(tasks: List<Task>, canvas: Canvas) {
+    val taskBarOffset = (input.rowHeight - input.barHeight) / 2
+    val taskActivityScene = TaskActivitySceneBuilder(
+      ActivitySceneTaskApi, input, canvas, TaskLabelSceneBuilderStub, TaskActivitySceneBuilder.Style(taskBarOffset)
+    )
+    tasks.forEachIndexed { row, task ->
+      taskActivityScene.renderActivities(row, listOf(task.activity), input.bottomUnitOffsets)
+      val nextLineY = row * input.rowHeight
+      val nextLine = canvas.createLine(0, nextLineY, input.viewportWidth, nextLineY)
+      nextLine.foregroundColor = Color.GRAY
+    }
+  }
+
+  private fun paintDependencies(tasks: List<Task>, canvas: Canvas) {
+    val dependencyRenderer = DependencySceneBuilder(
+      canvas, canvas.newLayer(), DependencySceneTaskApi(tasks, logger), { input.barHeight }
+    )
+    dependencyRenderer.build()
+  }
+
+  abstract class Task {
+    abstract val start: Date
+    abstract val end: Date
+    abstract val isMilestone: Boolean
+    abstract val subtasks: List<Task>
+    abstract val dependants: List<Dependant>
+
+    val activity = Activity()
+
+    fun flatten(result: MutableList<Task> = mutableListOf()): List<Task> {
+      result.add(this)
+      subtasks.forEach { it.flatten(result) }
+      return result
+    }
+
+    inner class Activity : BarChartActivity<Task> {
+      override fun getStart() = this@Task.start
+      override fun getEnd() = this@Task.end
+      override fun getOwner() = this@Task
+
+      override fun getDuration(): TimeDuration {
+        return TimeDurationImpl(
+          GPTimeUnitStack.DAY,
+          TimeUnit.DAYS.convert(start.time - end.time, TimeUnit.MILLISECONDS)
```

Comment:
I don't think that we want to reinvent duration building. There is `TaskManager::createDuration` method and I think it makes sense to get duration from there, via `Task` interface.

Reply:
I think `Task` interface requires too much to implement it here.